velox/connectors/hive/HiveDataSink.cpp

```diff
@@ -17,41 +17,93 @@
 #include "velox/connectors/hive/HiveDataSink.h"
 
 #include "velox/common/base/Fs.h"
+#include "velox/connectors/hive/HiveConfig.h"
 #include "velox/connectors/hive/HiveConnector.h"
-#include "velox/connectors/hive/HiveWriteProtocol.h"
 #include "velox/dwio/dwrf/writer/Writer.h"
 
+#include <boost/lexical_cast.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
 using namespace facebook::velox::dwrf;
 using WriterConfig = facebook::velox::dwrf::Config;
 
 namespace facebook::velox::connector::hive {
 
+namespace {
+
+std::string makeUuid() {
+  return boost::lexical_cast<std::string>(boost::uuids::random_generator()());
+}
+
+std::string makePartitionDirectory(
+    const std::string& tableDirectory,
+    const std::optional<std::string>& partitionSubdirectory) {
+  if (partitionSubdirectory.has_value()) {
+    return (fs::path(tableDirectory) / partitionSubdirectory.value()).string();
+  }
+  return tableDirectory;
+}
+
+} // namespace
+
 HiveDataSink::HiveDataSink(
     RowTypePtr inputType,
     std::shared_ptr<const HiveInsertTableHandle> insertTableHandle,
-    const ConnectorQueryCtx* FOLLY_NONNULL connectorQueryCtx,
-    std::shared_ptr<WriteProtocol> writeProtocol)
+    const ConnectorQueryCtx* connectorQueryCtx,
+    CommitStrategy commitStrategy)
     : inputType_(std::move(inputType)),
       insertTableHandle_(std::move(insertTableHandle)),
       connectorQueryCtx_(connectorQueryCtx),
-      writeProtocol_(std::move(writeProtocol)) {
-  VELOX_CHECK_NOT_NULL(
-      writeProtocol_, "Write protocol could not be nullptr for HiveDataSink.");
-}
-
-std::shared_ptr<ConnectorCommitInfo> HiveDataSink::getConnectorCommitInfo()
-    const {
-  return std::make_shared<HiveConnectorCommitInfo>(writerParameters_);
-}
+      commitStrategy_(commitStrategy) {}
 
 void HiveDataSink::appendData(VectorPtr input) {
   // For the time being the hive data sink supports one file
   // To extend it we can create a new writer for every
   // partition
   if (writers_.empty()) {
-    writers_.emplace_back(createWriter());
+    std::shared_ptr<const HiveWriterParameters> hiveWriterParameters =
+        getWriterParameters(std::nullopt);
+    writers_.emplace_back(createWriter(hiveWriterParameters));
+    writerInfo_.emplace_back(
+        std::make_shared<HiveWriterInfo>(*hiveWriterParameters));
   }
   writers_[0]->write(input);
+  writerInfo_[0]->numWrittenRows += input->size();
+}
+
+std::vector<std::string> HiveDataSink::finish() const {
+  std::vector<std::string> partitionUpdates;
+  partitionUpdates.reserve(writerInfo_.size());
+
+  for (const auto info : writerInfo_) {
+    if (info) {
```

Comment:
Yeah, I think VectorHasher reserves 0 for all-null value/partition value. If the input doesn't have the all-null partition, then the writeInfo_ as well as the writers_ will be set to nullptr at index 0.

(I know it's not being used in this PR. Just do the check here in case I forget later.)